python/kserve/kserve/logging.py

```diff
@@ -0,0 +1,51 @@
+# Copyright 2022 The KServe Authors.
```

Comment:
Hi @cmaddalozzo, I hope you don't mind me jumping into this PR. We rely extensively on KServe in our applications and would like to share some knowledge about how to configure loggers for libraries to ensure that the library's loggers do not interfere with application code.

First, please note that a library should ideally not be configuring its own logging, and definitely not be adding its own handlers. The "Note" warnings in the documentation explain the reasoning: https://docs.python.org/3/howto/logging.html#configuring-logging-for-a-library This was the cause of the original problem where KServe's logging configuration was taking over the logging configuration of our application code, which caused our logs to disappear. (Some of our logs are monitored for critical system health, so we rely on them extensively.)

Instead, the way it's typically done is shown in this [StackOverflow post](https://stackoverflow.com/questions/15727420/using-logging-in-multiple-modules).

Concretely, the library code defines module-level loggers like this:
```python
# my_library_class.py
import logging
# Note that no logging configuration is done here.
# This is critically important. This logger will inherit its configuration from the application code's configuration.
logger = logging.getLogger(__name__)

class MyLibraryClass:
    def work(self) -> None:
        logger.info("Logging message")
```
and the application code defines the logging configuration:
```python
# main.py
import logging

def main() -> None:
    start_application()

if __name__ == "__main__":
    # This is where the application code will define its own logging configuration
    # This includes the logging configuration for any libraries that it uses
    logging.basicConfig(....)
    main()
```

The way this PR is currently written, we have this line:
```python
from .logging import logger
```
that is causing all of this logging configuration to be executed when `ModelServer` is imported into application code. This is what causes problems for downstream application developers, because they do not expect a library that they import to change their existing logging configuration. The same concept applies elsewhere; for example, there is more logging configuration done [here](https://github.com/kserve/kserve/blob/422997007d10cca79dd142a6af4292853f530fa1/python/kserve/kserve/protocol/rest/server.py#L140) that also affects downstream application developers.

Of course, sometimes it is necessary for a library to configure a logger for debugging or testing purposes. That is fine, but when that is necessary, the logging configuration should be kept entirely separate from the code that is exposed to downstream application developers.

Reply:
Thanks @cmaddalozzo for the reply! I think there might be 2 ways to move forward following different conventions:

1. Do not setup logging inside the library
- Where we can have `setup_logging` function that takes a `log_config` as a `dict` which has the default value we have here.

2. Follow the `uvicorn` or `flask` server paradigm to 1)setup logging in lib and 2)allow user to pass `log_config` to the server instance to configure for logging. Indeed they configure logging in the library & they have provided easy way to configure custom logging.
- We can also change kserve's `ModelServer` and [`UvicornServer`](https://github.com/kserve/kserve/blob/master/python/kserve/kserve/protocol/rest/server.py#L140) to be like the original [`UvicornServer`](https://github.com/encode/uvicorn/blob/master/uvicorn/config.py#L82) where a default `log_config` is provided but it is easily overwriteable. 
 
Currently the kserve's `ModelServer`'s [`UvicornServer`](https://github.com/kserve/kserve/blob/master/python/kserve/kserve/model_server.py#L170) has no way to overwrite the `log_config` with a `dict`. Also the current default `log_config` file is wrong without [this line](https://github.com/encode/uvicorn/blob/master/uvicorn/config.py#L84).

Reply:
Thanks @cmaddalozzo for making the change! If I understand correctly that now we default to configure the logging by `KSERVE_LOG_CONFIG` and provide options for developer to either disable log_config or passing log_config of their own. I saw that is consistent with what the `UvicornServer` does. And it makes sense to me.

Reply:
hey @Yucao42, that's correct. We are basically relying on Uvicorn to configure all logging (both its own and KServe's) while giving the developer the option to disable this behavior by passing `log_config=None` or provide their own config by passing `log_config='path_to_log_config.[yaml|json]'` or `log_config=log_config_dict`. 

I don't love relying on Uvicorn to configure logging but it's convenient since they allow providing the config as either a path to the log config or a dictionary containing the config. One downside is that all logs output by the KServe logger will have the default configuration until the Uvicorn server is instantiated and the logging is configured.

Reply:
@Yucao42 hi! All the conversation makes sense, I am wondering if you could give me your opinion about the recent work that was added to allow Kserve to have a good access log format. Due to limitations in Uvicorn and ASGI specs (see https://github.com/encode/uvicorn/pull/947#issuecomment-1283006367) I added a specific option to use asgi-logger and be able to add fields like User Agent etc.. to the access log format (otherwise it seems not possible with the default Uvicorn's config). The library creates its own handler and is able to allow a custom format like Gunicorn's one. The complete change is https://github.com/kserve/kserve/pull/2782. Do you think it is ok to keep it? Or do you have an alternative suggestion about how to customize the access logs? Thanks!

Reply:
@elukey Hello! I skimmed over the PR. I think there are two things that are fixed in this PR: 1) the [default config value ](https://github.com/kserve/kserve/pull/2782/files#diff-238df0559a45edddf58f285b6faa318bdb72012bdc89f3f5433916ddfe7b7685R140-R181) disables all other loggings, 2) we could preserve the `UvicornServer`'s API of passing config of our own (either `Dict` or `str` as a json file). How do we do 2) might be a little discussion: currently we add another variable of `log_config` as `Dict`. I advocate for using `log_config` as both types like `UvicornServer` [here](https://github.com/encode/uvicorn/blob/d43afed1cfa018a85c83094da8a2dd29f656d676/uvicorn/config.py#L212), but current code(adding `logging_config`) in this PR makes sense too. I appreciate both your efforts! @cmaddalozzo @elukey !